File: python/strategy.py
"""
Trading strategy and backtesting utilities for Multi-Scale Attention.

This module provides tools for backtesting trading strategies based on
multi-scale attention predictions.
"""

from dataclasses import dataclass
from typing import Dict, List, Optional, Tuple

import numpy as np
import pandas as pd
import torch


@dataclass
class BacktestResult:
    """Container for backtest results."""

    total_return: float
    sharpe_ratio: float
    sortino_ratio: float
    max_drawdown: float
    win_rate: float
    profit_factor: float
    n_trades: int
    avg_trade_return: float
    volatility: float
    calmar_ratio: float
    equity_curve: pd.Series
    trades: pd.DataFrame


def backtest_multi_scale_strategy(
    predictions: Dict[str, np.ndarray],
    prices: pd.DataFrame,
    initial_capital: float = 100000,
    transaction_cost: float = 0.001,
    confidence_threshold: float = 0.55,
    position_sizing: str = "confidence",
    max_position: float = 1.0,
    stop_loss: Optional[float] = None,
    take_profit: Optional[float] = None,
) -> BacktestResult:
    """
    Backtest a multi-scale attention trading strategy.

    Args:
        predictions: Model predictions dictionary with keys:
            - 'direction': Direction probabilities [n_samples, 1]
            - 'short_term': Short-term predictions [n_samples, 1]
            - 'uncertainty': Uncertainty estimates [n_samples, 1]
        prices: DataFrame with 'close' column
        initial_capital: Starting capital
        transaction_cost: Cost per transaction (fraction)
        confidence_threshold: Minimum direction confidence to trade
        position_sizing: 'fixed', 'confidence', or 'kelly'
        max_position: Maximum position size (fraction of capital)
        stop_loss: Stop loss threshold (fraction)
        take_profit: Take profit threshold (fraction)

    Returns:
        BacktestResult with performance metrics
    """
    n_samples = len(predictions["direction"])
    prices = prices.iloc[:n_samples].copy()

    capital = initial_capital
    position = 0.0
    entry_price = 0.0
    equity = [capital]
    returns = []
    trades = []

    for i in range(n_samples - 1):
        current_price = prices["close"].iloc[i]
        next_price = prices["close"].iloc[i + 1]

        # Get signals
        direction_prob = predictions["direction"][i, 0]
        uncertainty = (
            predictions.get("uncertainty", np.ones_like(predictions["direction"]))[i, 0]
        )

        # Determine confidence
        confidence = abs(direction_prob - 0.5) * 2

        # Check stop loss / take profit
        if position != 0 and entry_price > 0:
            unrealized_return = (current_price - entry_price) / entry_price * np.sign(
                position
            )

            if stop_loss is not None and unrealized_return < -stop_loss:
                # Stop loss triggered
                position_change = -position
                cost = abs(position_change) * transaction_cost * capital
                capital -= cost

                trades.append(
                    {
                        "index": i,
                        "type": "stop_loss",
                        "price": current_price,
                        "position": position,
                        "cost": cost,
                    }
                )
                position = 0
                entry_price = 0

            elif take_profit is not None and unrealized_return > take_profit:
                # Take profit triggered
                position_change = -position
                cost = abs(position_change) * transaction_cost * capital
                capital -= cost

                trades.append(
                    {
                        "index": i,
                        "type": "take_profit",
                        "price": current_price,
                        "position": position,
                        "cost": cost,
                    }
                )
                position = 0
                entry_price = 0

        # Determine target position
        if confidence > confidence_threshold:
            if position_sizing == "fixed":
                target_position = max_position if direction_prob > 0.5 else -max_position
            elif position_sizing == "confidence":
                target_position = confidence * max_position
                if direction_prob < 0.5:
                    target_position = -target_position
            elif position_sizing == "kelly":
                # Simplified Kelly criterion
                win_prob = direction_prob if direction_prob > 0.5 else (1 - direction_prob)
                kelly_fraction = (2 * win_prob - 1) / 1  # Assuming 1:1 payoff
                kelly_fraction = max(0, min(kelly_fraction, 0.25))  # Cap at 25%
                target_position = kelly_fraction * max_position
                if direction_prob < 0.5:
                    target_position = -target_position
            else:
                target_position = 0.0

            # Adjust for uncertainty (reduce position if uncertain)
            if uncertainty > 0:
                target_position = target_position / (1 + uncertainty)
        else:
            target_position = 0.0

        # Calculate position change
        position_change = target_position - position

        # Transaction costs
        if abs(position_change) > 0.01:
            cost = abs(position_change) * transaction_cost * capital
            capital -= cost

            if position_change != 0:
                entry_price = current_price

            trades.append(
                {
                    "index": i,
                    "type": "long" if target_position > 0 else "short",
                    "price": current_price,
                    "position": target_position,
                    "cost": cost,
                }
            )

        position = target_position

        # Calculate return
        price_return = (next_price - current_price) / current_price
        pnl = position * price_return * capital
        capital += pnl

        if capital > 0:
            returns.append(pnl / equity[-1])
        else:
            returns.append(0)

        equity.append(capital)

    # Calculate metrics
    equity_series = pd.Series(equity)
    returns_series = pd.Series(returns)

    total_return = (capital - initial_capital) / initial_capital

    # Annualized Sharpe ratio (assuming hourly data)
    periods_per_year = 24 * 365
    if len(returns_series) > 1 and returns_series.std() > 0:
        sharpe = returns_series.mean() / returns_series.std() * np.sqrt(periods_per_year)
    else:
        sharpe = 0

    # Sortino ratio
    downside_returns = returns_series[returns_series < 0]
    if len(downside_returns) > 0 and downside_returns.std() > 0:
        sortino = (
            returns_series.mean() / downside_returns.std() * np.sqrt(periods_per_year)
        )
    else:
        sortino = 0

    # Max drawdown
    running_max = equity_series.cummax()
    drawdown = (equity_series - running_max) / running_max
    max_drawdown = drawdown.min()

    # Win rate
    winning_returns = returns_series[returns_series > 0]
    win_rate = len(winning_returns) / len(returns_series) if len(returns_series) > 0 else 0

    # Profit factor
    gross_profit = winning_returns.sum() if len(winning_returns) > 0 else 0
    losing_returns = returns_series[returns_series < 0]
    gross_loss = abs(losing_returns.sum()) if len(losing_returns) > 0 else 0
    profit_factor = gross_profit / gross_loss if gross_loss > 0 else float("inf")

    # Volatility (annualized)
    volatility = returns_series.std() * np.sqrt(periods_per_year)

    # Calmar ratio
    calmar_ratio = total_return / abs(max_drawdown) if max_drawdown < 0 else 0

    # Average trade return
    avg_trade_return = returns_series.mean() if len(returns_series) > 0 else 0

    return BacktestResult(
        total_return=total_return,
        sharpe_ratio=sharpe,
        sortino_ratio=sortino,
        max_drawdown=max_drawdown,
        win_rate=win_rate,
        profit_factor=profit_factor,
        n_trades=len(trades),
        avg_trade_return=avg_trade_return,
        volatility=volatility,
        calmar_ratio=calmar_ratio,
        equity_curve=equity_series,
        trades=pd.DataFrame(trades) if trades else pd.DataFrame(),
    )


def compare_scale_strategies(
    predictions: Dict[str, np.ndarray],
    prices: pd.DataFrame,
    **backtest_kwargs,
) -> pd.DataFrame:
    """
    Compare strategies using different horizon predictions.

    Args:
        predictions: Model predictions dictionary
        prices: DataFrame with price data
        **backtest_kwargs: Additional arguments for backtest_multi_scale_strategy

    Returns:
        DataFrame comparing different strategies
    """
    results = {}

    horizons = ["short_term", "medium_term", "long_term"]

    for horizon in horizons:
        if horizon not in predictions:
            continue

        # Create predictions using only this horizon for direction
        single_horizon_preds = {
            "direction": (predictions[horizon] > 0).astype(float),
            "short_term": predictions[horizon],
        }

        if "uncertainty" in predictions:
            single_horizon_preds["uncertainty"] = predictions["uncertainty"]

        result = backtest_multi_scale_strategy(
            single_horizon_preds, prices, **backtest_kwargs
        )

        results[horizon] = {
            "Total Return": f"{result.total_return * 100:.2f}%",
            "Sharpe Ratio": f"{result.sharpe_ratio:.2f}",
            "Sortino Ratio": f"{result.sortino_ratio:.2f}",
            "Max Drawdown": f"{result.max_drawdown * 100:.2f}%",
            "Win Rate": f"{result.win_rate * 100:.1f}%",
            "Profit Factor": f"{result.profit_factor:.2f}",
            "Trades": result.n_trades,
            "Calmar Ratio": f"{result.calmar_ratio:.2f}",
        }

    # Add combined strategy
    result = backtest_multi_scale_strategy(predictions, prices, **backtest_kwargs)
    results["multi_scale"] = {
        "Total Return": f"{result.total_return * 100:.2f}%",
        "Sharpe Ratio": f"{result.sharpe_ratio:.2f}",
        "Sortino Ratio": f"{result.sortino_ratio:.2f}",
        "Max Drawdown": f"{result.max_drawdown * 100:.2f}%",
        "Win Rate": f"{result.win_rate * 100:.1f}%",
        "Profit Factor": f"{result.profit_factor:.2f}",
        "Trades": result.n_trades,
        "Calmar Ratio": f"{result.calmar_ratio:.2f}",
    }

    return pd.DataFrame(results).T


def calculate_information_ratio(
    strategy_returns: np.ndarray,
    benchmark_returns: np.ndarray,
    periods_per_year: int = 24 * 365,
) -> float:
    """
    Calculate the Information Ratio.

    Args:
        strategy_returns: Strategy returns array
        benchmark_returns: Benchmark returns array
        periods_per_year: Number of periods per year

    Returns:
        Information Ratio
    """
    excess_returns = strategy_returns - benchmark_returns
    tracking_error = np.std(excess_returns)

    if tracking_error == 0:
        return 0

    return np.mean(excess_returns) / tracking_error * np.sqrt(periods_per_year)


def calculate_var(
    returns: np.ndarray, confidence: float = 0.95, method: str = "historical"
) -> float:
    """
    Calculate Value at Risk.

    Args:
        returns: Returns array
        confidence: Confidence level
        method: 'historical' or 'parametric'

    Returns:
        VaR value (negative number representing loss)
    """
    if method == "historical":
        return np.percentile(returns, (1 - confidence) * 100)
    elif method == "parametric":
        from scipy import stats

        mean = np.mean(returns)
        std = np.std(returns)
        return mean + std * stats.norm.ppf(1 - confidence)
    else:
        raise ValueError(f"Unknown method: {method}")


def calculate_expected_shortfall(
    returns: np.ndarray, confidence: float = 0.95
) -> float:
    """
    Calculate Expected Shortfall (Conditional VaR).

    Args:
        returns: Returns array
        confidence: Confidence level

    Returns:
        Expected Shortfall value
    """
    var = calculate_var(returns, confidence)
    return np.mean(returns[returns <= var])


def generate_performance_report(result: BacktestResult) -> str:
    """
    Generate a text performance report.

    Args:
        result: BacktestResult from backtesting

    Returns:
        Formatted performance report string
    """
    report = """
================================================================================
                        MULTI-SCALE ATTENTION STRATEGY REPORT
================================================================================

PERFORMANCE METRICS
-------------------
Total Return:        {total_return:>10.2%}
Sharpe Ratio:        {sharpe:>10.2f}
Sortino Ratio:       {sortino:>10.2f}
Calmar Ratio:        {calmar:>10.2f}

RISK METRICS
------------
Maximum Drawdown:    {max_dd:>10.2%}
Volatility (Ann.):   {volatility:>10.2%}

TRADING METRICS
---------------
Total Trades:        {n_trades:>10d}
Win Rate:            {win_rate:>10.2%}
Profit Factor:       {profit_factor:>10.2f}
Avg Trade Return:    {avg_trade:>10.4%}

================================================================================
    """.format(
        total_return=result.total_return,
        sharpe=result.sharpe_ratio,
        sortino=result.sortino_ratio,
        calmar=result.calmar_ratio,
        max_dd=result.max_drawdown,
        volatility=result.volatility,
        n_trades=result.n_trades,
        win_rate=result.win_rate,
        profit_factor=result.profit_factor,
        avg_trade=result.avg_trade_return,
    )

    return report


def plot_equity_curve(
    result: BacktestResult,
    title: str = "Multi-Scale Attention Strategy",
    save_path: Optional[str] = None,
) -> None:
    """
    Plot the equity curve.

    Args:
        result: BacktestResult from backtesting
        title: Plot title
        save_path: Optional path to save the plot
    """
    try:
        import matplotlib.pyplot as plt
    except ImportError:
        print("matplotlib not installed, skipping plot")
        return

    fig, axes = plt.subplots(2, 1, figsize=(12, 8), sharex=True)

    # Equity curve
    axes[0].plot(result.equity_curve, color="steelblue", linewidth=1)
    axes[0].fill_between(
        range(len(result.equity_curve)),
        result.equity_curve.iloc[0],
        result.equity_curve,
        alpha=0.3,
    )
    axes[0].set_title(title)
    axes[0].set_ylabel("Portfolio Value")
    axes[0].grid(True, alpha=0.3)

    # Drawdown
    running_max = result.equity_curve.cummax()
    drawdown = (result.equity_curve - running_max) / running_max
    axes[1].fill_between(
        range(len(drawdown)), 0, drawdown * 100, color="red", alpha=0.5
    )
    axes[1].set_ylabel("Drawdown (%)")
    axes[1].set_xlabel("Time")
    axes[1].grid(True, alpha=0.3)

    plt.tight_layout()

    if save_path:
        plt.savefig(save_path, dpi=150)
    else:
        plt.show()

    plt.close()

File: rust/src/strategy.rs
//! Backtesting and strategy evaluation for multi-scale attention.
//!
//! This module provides utilities for:
//! - Backtesting trading strategies
//! - Performance metrics calculation
//! - Strategy comparison

use crate::model::Predictions;
use std::collections::HashMap;

/// Configuration for backtesting
#[derive(Debug, Clone)]
pub struct BacktestConfig {
    /// Initial capital
    pub initial_capital: f64,
    /// Transaction cost (as fraction, e.g., 0.001 = 0.1%)
    pub transaction_cost: f64,
    /// Confidence threshold for trading
    pub confidence_threshold: f64,
    /// Maximum position size (as fraction of capital)
    pub max_position_size: f64,
    /// Stop loss threshold (as fraction)
    pub stop_loss: Option<f64>,
    /// Take profit threshold (as fraction)
    pub take_profit: Option<f64>,
}

impl Default for BacktestConfig {
    fn default() -> Self {
        Self {
            initial_capital: 100_000.0,
            transaction_cost: 0.001,
            confidence_threshold: 0.55,
            max_position_size: 1.0,
            stop_loss: None,
            take_profit: None,
        }
    }
}

/// Result of backtesting
#[derive(Debug, Clone)]
pub struct BacktestResult {
    /// Equity curve over time
    pub equity_curve: Vec<f64>,
    /// Returns for each period
    pub returns: Vec<f64>,
    /// Positions taken (-1, 0, 1)
    pub positions: Vec<i32>,
    /// Trade signals
    pub signals: Vec<i32>,
    /// Total return
    pub total_return: f64,
    /// Annualized return
    pub annualized_return: f64,
    /// Sharpe ratio
    pub sharpe_ratio: f64,
    /// Sortino ratio
    pub sortino_ratio: f64,
    /// Maximum drawdown
    pub max_drawdown: f64,
    /// Win rate
    pub win_rate: f64,
    /// Number of trades
    pub n_trades: usize,
    /// Profit factor
    pub profit_factor: f64,
    /// Calmar ratio
    pub calmar_ratio: f64,
}

impl BacktestResult {
    /// Generate a performance report string
    pub fn report(&self) -> String {
        format!(
            r#"
================================================================================
                        MULTI-SCALE STRATEGY BACKTEST RESULTS
================================================================================

PERFORMANCE METRICS
------------------
Total Return:        {:>10.2}%
Annualized Return:   {:>10.2}%
Sharpe Ratio:        {:>10.3}
Sortino Ratio:       {:>10.3}
Calmar Ratio:        {:>10.3}
Max Drawdown:        {:>10.2}%

TRADING STATISTICS
------------------
Number of Trades:    {:>10}
Win Rate:            {:>10.2}%
Profit Factor:       {:>10.3}

EQUITY
------
Initial Capital:     {:>10.2}
Final Equity:        {:>10.2}

================================================================================
"#,
            self.total_return * 100.0,
            self.annualized_return * 100.0,
            self.sharpe_ratio,
            self.sortino_ratio,
            self.calmar_ratio,
            self.max_drawdown * 100.0,
            self.n_trades,
            self.win_rate * 100.0,
            self.profit_factor,
            self.equity_curve.first().unwrap_or(&0.0),
            self.equity_curve.last().unwrap_or(&0.0),
        )
    }
}

/// Backtest a multi-scale strategy
///
/// # Arguments
/// * `predictions` - Model predictions
/// * `prices` - Price series for backtesting
/// * `config` - Backtest configuration
///
/// # Returns
/// BacktestResult with performance metrics
pub fn backtest_multi_scale_strategy(
    predictions: &Predictions,
    prices: &[f64],
    config: &BacktestConfig,
) -> BacktestResult {
    let n = predictions.direction.nrows().min(prices.len());

    if n == 0 {
        return BacktestResult {
            equity_curve: vec![config.initial_capital],
            returns: vec![],
            positions: vec![],
            signals: vec![],
            total_return: 0.0,
            annualized_return: 0.0,
            sharpe_ratio: 0.0,
            sortino_ratio: 0.0,
            max_drawdown: 0.0,
            win_rate: 0.0,
            n_trades: 0,
            profit_factor: 0.0,
            calmar_ratio: 0.0,
        };
    }

    let mut equity = config.initial_capital;
    let mut equity_curve = vec![equity];
    let mut returns = Vec::with_capacity(n);
    let mut positions = Vec::with_capacity(n);
    let mut signals = Vec::with_capacity(n);
    let mut current_position = 0i32;
    let mut n_trades = 0;
    let mut wins = 0;
    let mut gross_profit = 0.0;
    let mut gross_loss = 0.0;

    for i in 0..n - 1 {
        // Get prediction confidence
        let direction_prob = predictions.direction[[i, 0]];
        let uncertainty = predictions.uncertainty[[i, 0]];

        // Generate signal based on confidence
        let signal = if direction_prob > config.confidence_threshold && uncertainty < 0.5 {
            1 // Long
        } else if direction_prob < (1.0 - config.confidence_threshold) && uncertainty < 0.5 {
            -1 // Short
        } else {
            0 // Neutral
        };

        signals.push(signal);

        // Calculate position change
        let position_change = signal - current_position;

        // Apply transaction costs for position changes
        if position_change != 0 {
            let cost = equity * config.transaction_cost * position_change.abs() as f64;
            equity -= cost;
            n_trades += 1;
        }

        // Calculate return for this period
        let price_return = if i + 1 < prices.len() && prices[i] > 0.0 {
            (prices[i + 1] - prices[i]) / prices[i]
        } else {
            0.0
        };

        // Apply position return
        let position_return = current_position as f64 * price_return * config.max_position_size;
        let period_pnl = equity * position_return;

        // Track wins/losses
        if position_return > 0.0 {
            wins += 1;
            gross_profit += period_pnl;
        } else if position_return < 0.0 {
            gross_loss += period_pnl.abs();
        }

        equity += period_pnl;
        returns.push(position_return);
        positions.push(current_position);
        equity_curve.push(equity);

        current_position = signal;
    }

    // Add final position
    positions.push(current_position);
    signals.push(0);

    // Calculate metrics
    let total_return = (equity - config.initial_capital) / config.initial_capital;

    // Annualized return (assuming daily data, 252 trading days)
    let n_periods = returns.len() as f64;
    let annualized_return = if n_periods > 0.0 {
        ((1.0 + total_return).powf(252.0 / n_periods)) - 1.0
    } else {
        0.0
    };

    // Calculate Sharpe ratio
    let mean_return = if !returns.is_empty() {
        returns.iter().sum::<f64>() / returns.len() as f64
    } else {
        0.0
    };

    let std_return = if returns.len() > 1 {
        let variance = returns
            .iter()
            .map(|r| (r - mean_return).powi(2))
            .sum::<f64>()
            / (returns.len() - 1) as f64;
        variance.sqrt()
    } else {
        0.0
    };

    let sharpe_ratio = if std_return > 0.0 {
        (mean_return * 252.0_f64.sqrt()) / std_return
    } else {
        0.0
    };

    // Calculate Sortino ratio (downside deviation)
    let downside_returns: Vec<f64> = returns.iter().filter(|&&r| r < 0.0).copied().collect();
    let downside_std = if downside_returns.len() > 1 {
        let variance = downside_returns
            .iter()
            .map(|r| r.powi(2))
            .sum::<f64>()
            / downside_returns.len() as f64;
        variance.sqrt()
    } else {
        0.0
    };

    let sortino_ratio = if downside_std > 0.0 {
        (mean_return * 252.0_f64.sqrt()) / downside_std
    } else {
        0.0
    };

    // Calculate maximum drawdown
    let mut peak = equity_curve[0];
    let mut max_drawdown = 0.0;
    for &equity in &equity_curve {
        if equity > peak {
            peak = equity;
        }
        let drawdown = (peak - equity) / peak;
        if drawdown > max_drawdown {
            max_drawdown = drawdown;
        }
    }

    // Win rate
    let win_rate = if n_trades > 0 {
        wins as f64 / n_trades as f64
    } else {
        0.0
    };

    // Profit factor
    let profit_factor = if gross_loss > 0.0 {
        gross_profit / gross_loss
    } else if gross_profit > 0.0 {
        f64::INFINITY
    } else {
        0.0
    };

    // Calmar ratio
    let calmar_ratio = if max_drawdown > 0.0 {
        annualized_return / max_drawdown
    } else if annualized_return > 0.0 {
        f64::INFINITY
    } else {
        0.0
    };

    BacktestResult {
        equity_curve,
        returns,
        positions,
        signals,
        total_return,
        annualized_return,
        sharpe_ratio,
        sortino_ratio,
        max_drawdown,
        win_rate,
        n_trades,
        profit_factor,
        calmar_ratio,
    }
}

/// Compare strategies using different scales
pub fn compare_scale_strategies(
    predictions: &Predictions,
    prices: &[f64],
    config: &BacktestConfig,
) -> HashMap<String, BacktestResult> {
    let mut results = HashMap::new();

    // Full multi-scale strategy
    let multi_scale_result = backtest_multi_scale_strategy(predictions, prices, config);
    results.insert("multi_scale".to_string(), multi_scale_result);

    // Short-term only strategy (using short-term predictions for direction)
    let short_term_predictions = Predictions {
        short_term: predictions.short_term.clone(),
        medium_term: predictions.short_term.clone(),
        long_term: predictions.short_term.clone(),
        direction: predictions.short_term.mapv(|v| if v > 0.0 { 0.7 } else { 0.3 }),
        uncertainty: predictions.uncertainty.clone(),
    };
    let short_term_result = backtest_multi_scale_strategy(&short_term_predictions, prices, config);
    results.insert("short_term_only".to_string(), short_term_result);

    // Long-term only strategy
    let long_term_predictions = Predictions {
        short_term: predictions.long_term.clone(),
        medium_term: predictions.long_term.clone(),
        long_term: predictions.long_term.clone(),
        direction: predictions.long_term.mapv(|v| if v > 0.0 { 0.7 } else { 0.3 }),
        uncertainty: predictions.uncertainty.clone(),
    };
    let long_term_result = backtest_multi_scale_strategy(&long_term_predictions, prices, config);
    results.insert("long_term_only".to_string(), long_term_result);

    results
}

/// Calculate Value at Risk (VaR) at given confidence level
pub fn calculate_var(returns: &[f64], confidence: f64) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }

    let mut sorted_returns: Vec<f64> = returns.to_vec();
    sorted_returns.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));

    let idx = ((1.0 - confidence) * sorted_returns.len() as f64) as usize;
    let idx = idx.min(sorted_returns.len() - 1);

    -sorted_returns[idx]
}

/// Calculate Expected Shortfall (CVaR) at given confidence level
pub fn calculate_expected_shortfall(returns: &[f64], confidence: f64) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }

    let mut sorted_returns: Vec<f64> = returns.to_vec();
    sorted_returns.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));

    let cutoff_idx = ((1.0 - confidence) * sorted_returns.len() as f64) as usize;
    let cutoff_idx = cutoff_idx.max(1);

    let tail: Vec<f64> = sorted_returns.iter().take(cutoff_idx).copied().collect();
    let es: f64 = tail.iter().sum::<f64>() / tail.len() as f64;

    -es
}

/// Generate comparison report for multiple strategies
pub fn generate_comparison_report(results: &HashMap<String, BacktestResult>) -> String {
    let mut report = String::new();
    report.push_str("\n================================================================================\n");
    report.push_str("                     STRATEGY COMPARISON REPORT\n");
    report.push_str("================================================================================\n\n");

    report.push_str(&format!(
        "{:<20} {:>12} {:>12} {:>12} {:>12}\n",
        "Strategy", "Return %", "Sharpe", "Sortino", "Max DD %"
    ));
    report.push_str(&format!("{}\n", "-".repeat(72)));

    for (name, result) in results {
        report.push_str(&format!(
            "{:<20} {:>12.2} {:>12.3} {:>12.3} {:>12.2}\n",
            name,
            result.total_return * 100.0,
            result.sharpe_ratio,
            result.sortino_ratio,
            result.max_drawdown * 100.0
        ));
    }

    report.push_str(&format!("{}\n", "-".repeat(72)));
    report.push_str("\n================================================================================\n");

    report
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::Array2;

    fn create_test_predictions(n: usize) -> Predictions {
        Predictions {
            short_term: Array2::from_elem((n, 1), 0.01),
            medium_term: Array2::from_elem((n, 1), 0.02),
            long_term: Array2::from_elem((n, 1), 0.03),
            direction: Array2::from_elem((n, 1), 0.6),
            uncertainty: Array2::from_elem((n, 1), 0.1),
        }
    }

    #[test]
    fn test_backtest_basic() {
        let predictions = create_test_predictions(100);
        let prices: Vec<f64> = (0..100)
            .map(|i| 100.0 * (1.0 + 0.001 * i as f64))
            .collect();

        let config = BacktestConfig::default();
        let result = backtest_multi_scale_strategy(&predictions, &prices, &config);

        assert!(!result.equity_curve.is_empty());
        assert_eq!(result.positions.len(), 100);
    }

    #[test]
    fn test_var_calculation() {
        let returns = vec![-0.05, -0.03, -0.01, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07];
        let var_95 = calculate_var(&returns, 0.95);
        assert!(var_95 > 0.0);
    }

    #[test]
    fn test_expected_shortfall() {
        let returns = vec![-0.05, -0.03, -0.01, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07];
        let es = calculate_expected_shortfall(&returns, 0.95);
        assert!(es > 0.0);
    }

    #[test]
    fn test_compare_strategies() {
        let predictions = create_test_predictions(50);
        let prices: Vec<f64> = (0..50)
            .map(|i| 100.0 * (1.0 + 0.001 * i as f64))
            .collect();

        let config = BacktestConfig::default();
        let comparison = compare_scale_strategies(&predictions, &prices, &config);

        assert!(comparison.contains_key("multi_scale"));
        assert!(comparison.contains_key("short_term_only"));
        assert!(comparison.contains_key("long_term_only"));
    }
}
